File: 116-populating-next-right-pointers-in-each-node/populating-next-right-pointers-in-each-node.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}
    
    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
*/

class Solution {
    public Node connect(Node root) {
        if(root == null){
            return root;
        }
        root.next = null;
        fun(root);
        return root;
    }
    public void fun(Node node){
        if(node == null){
            return;
        }
        if(node.left != null){
            node.left.next = node.right;
        }
        if(node.right != null){
            if(node.next != null){
                node.right.next = node.next.left;
            }
            else{
                node.right.next = null;
            }
        }
        fun(node.left);
        fun(node.right);
    }


}

File: 116-populating-next-right-pointers-in-each-node/populating-next-right-pointers-in-each-node.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution:
    def connect(self, root: 'Optional[Node]') -> 'Optional[Node]':
        def fun(node,isLeft,parent):
            if node:
                if parent:
                    if isLeft:
                        node.next = parent.right
                    else:
                        if parent.next:
                            node.next = parent.next.left
                fun(node.left,True,node)
                fun(node.right,False,node)
        fun(root,True,None)
        return root
